src/promptflow/promptflow/_sdk/_record_storage.py

```diff
@@ -0,0 +1,228 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import collections
+import hashlib
+import json
+from pathlib import Path
+from typing import Any, Dict, OrderedDict
+
+from promptflow._sdk._configuration import Configuration
+from promptflow._sdk._constants import RecordMode
+from promptflow._sdk._errors import RecordFileMissingException, RecordItemMissingException
+from promptflow._utils.dataclass_serializer import serialize
+from promptflow._utils.tool_utils import get_inputs_for_prompt_template
+
+
+class RecordStorage(object):
+    """
+    RecordStorage is used to store the record of node run.
+    File often stored in .promptflow/node_cache.json
+    Example of cached items:
+    {
+        "/record/file/resolved": {
+            "hash_value": { # hash_value is sha1 of dict, accelerate the search
+                "input": {
+                    "key1": "value1", # Converted to string, type info dropped
+                },
+                "output": "output_convert_to_string",
+                "output_type": "output_type" # Currently support only simple types.
+            }
+        }
+    }
+    """
+
+    _standard_record_folder = ".promptflow"
+    _standard_record_name = "node_cache.json"
+    _instance = None
+
+    def __init__(self, record_file: str = None):
+        """
+        RecordStorage is used to store the record of node run.
+        """
+        self._record_file: Path = None
+        self.cached_items: Dict[str, Dict[str, Dict[str, object]]] = {}
+        self.record_file = record_file
+
+    @property
+    def record_file(self) -> Path:
+        return self._record_file
+
+    @record_file.setter
+    def record_file(self, record_file) -> None:
```

Comment:
add docstring for this indicating it will load file if exists

Reply:
OK